aptos-move/aptos-vm-types/src/storage/space_pricing.rs

```diff
@@ -127,8 +150,187 @@ impl DiskSpacePricing {
             },
             Deletion => ChargeAndRefund {
                 charge: 0.into(),
-                refund: metadata.total_deposit().into(),
+                refund: op.metadata_mut.total_deposit().into(),
+            },
+        }
+    }
+
+    fn charge_refund_write_op_v2(
+        params: &TransactionGasParameters,
+        op: WriteOpInfo,
+    ) -> ChargeAndRefund {
+        use WriteOpSize::*;
+
+        let key_size = op.key.size() as u64;
+        let num_bytes = key_size + op.op_size.write_len().unwrap_or(0);
+        let target_bytes_deposit: u64 = num_bytes * u64::from(params.storage_fee_per_state_byte);
+
+        match op.op_size {
+            Creation { .. } => {
+                // permanent storage fee
+                let slot_deposit = u64::from(params.storage_fee_per_state_slot);
+
+                op.metadata_mut.maybe_upgrade();
+                op.metadata_mut.set_slot_deposit(slot_deposit);
+                op.metadata_mut.set_bytes_deposit(target_bytes_deposit);
+
+                ChargeAndRefund {
+                    charge: (slot_deposit + target_bytes_deposit).into(),
+                    refund: 0.into(),
+                }
+            },
+            Modification { write_len } => {
```

Comment:
@igor-aptos @gelash How does this charging logic work with delayed fields? We will not be processing charges for modified resources with delayed fields, it seems?